File: dehaze-go/internal/service/favorite/favorite_service.go
package favorite

import (
	"context"
	"fmt"
	"time"

	"github.com/earthyzinc/dehaze-go/internal/model"
	"github.com/earthyzinc/dehaze-go/internal/model/bo"
	"github.com/earthyzinc/dehaze-go/internal/model/query"
	"github.com/earthyzinc/dehaze-go/internal/model/vo"
	algorepo "github.com/earthyzinc/dehaze-go/internal/repository/algorithm"
	datasetrepo "github.com/earthyzinc/dehaze-go/internal/repository/dataset"
	favrepo "github.com/earthyzinc/dehaze-go/internal/repository/favorite"
	memberrepo "github.com/earthyzinc/dehaze-go/internal/repository/member"
	predrepo "github.com/earthyzinc/dehaze-go/internal/repository/pred_log"
	dictservice "github.com/earthyzinc/dehaze-go/internal/service/dict"
	"github.com/earthyzinc/dehaze-go/pkg/common"
)

const (
	// favoriteCapacityDictType 收藏容量字典类型（sys_dict: favorite_capacity）
	favoriteCapacityDictType = "favorite_capacity"

	targetTypeAlgorithm = "algorithm"
	targetTypeResult    = "result"
	targetTypeDataset   = "dataset"
	targetTypeImage     = "image"
	targetTypePreset    = "preset"
)

var validTargetTypes = []string{targetTypeAlgorithm, targetTypeResult, targetTypeDataset, targetTypeImage, targetTypePreset}

type FavoriteService struct {
	favRepo       favrepo.IFavoriteRepository
	memberRepo    memberrepo.IMemberRepository
	algorithmRepo algorepo.IAlgorithmRepository
	predLogRepo   predrepo.IPredLogRepository
	datasetRepo   datasetrepo.IDatasetRepository
	dictSvc       dictservice.IDictService
}

func NewFavoriteService(
	favRepo favrepo.IFavoriteRepository,
	memberRepo memberrepo.IMemberRepository,
	algorithmRepo algorepo.IAlgorithmRepository,
	predLogRepo predrepo.IPredLogRepository,
	datasetRepo datasetrepo.IDatasetRepository,
	dictSvc dictservice.IDictService,
) *FavoriteService {
	return &FavoriteService{
		favRepo:       favRepo,
		memberRepo:    memberRepo,
		algorithmRepo: algorithmRepo,
		predLogRepo:   predLogRepo,
		datasetRepo:   datasetRepo,
		dictSvc:       dictSvc,
	}
}

func (s *FavoriteService) Add(ctx context.Context, userID int64, form *bo.FavoriteForm) (int64, error) {
	// 校验目标对象是否存在
	if err := s.checkTargetExists(ctx, form.TargetType, form.TargetID); err != nil {
		return 0, err
	}

	// 容量校验
	currentCount, err := s.favRepo.CountByUserID(ctx, userID)
	if err != nil {
		return 0, common.WrapBizError(common.DATABASE_ERROR, "查询收藏数量失败", err)
	}
	capacity := s.getCapacity(ctx, userID)
	if currentCount >= int64(capacity) {
		return 0, common.NewBizError(common.BUSINESS_ERROR,
			fmt.Sprintf("收藏已达上限（%d条），请清理后重试", capacity))
	}

	favorite := &model.SysFavorite{
		UserID:     userID,
		TargetType: form.TargetType,
		TargetID:   form.TargetID,
		IsInvalid:  0,
		Deleted:    0,
	}
	if err := s.favRepo.Upsert(ctx, favorite); err != nil {
		return 0, common.WrapBizError(common.DATABASE_ERROR, "添加收藏失败", err)
	}
	return favorite.ID, nil
}

func (s *FavoriteService) checkTargetExists(ctx context.Context, targetType string, targetID int64) error {
	switch targetType {
	case targetTypeAlgorithm:
		exists, err := s.algorithmRepo.ExistsByID(ctx, targetID)
		if err != nil {
			return common.WrapBizError(common.DATABASE_ERROR, "查询算法失败", err)
		}
		if !exists {
			return common.NewBizError(common.RESOURCE_NOT_FOUND, "算法不存在")
		}
	case targetTypeDataset:
		exists, err := s.datasetRepo.ExistsByID(ctx, targetID)
		if err != nil {
			return common.WrapBizError(common.DATABASE_ERROR, "查询数据集失败", err)
		}
		if !exists {
			return common.NewBizError(common.RESOURCE_NOT_FOUND, "数据集不存在")
		}
	case targetTypeResult:
		exists, err := s.predLogRepo.ExistsByID(ctx, targetID)
		if err != nil {
			return common.WrapBizError(common.DATABASE_ERROR, "查询处理记录失败", err)
		}
		if !exists {
			return common.NewBizError(common.RESOURCE_NOT_FOUND, "处理记录不存在")
		}
	}
	return nil
}

func (s *FavoriteService) DeleteByIDs(ctx context.Context, userID int64, ids []int64) error {
	if len(ids) == 0 {
		return nil
	}
	return s.favRepo.DeleteByIDs(ctx, userID, ids)
}

func (s *FavoriteService) GetPage(ctx context.Context, userID int64, q *query.FavoritePageQuery) (*vo.PageResult[vo.FavoriteVO], error) {
	rows, total, err := s.favRepo.FindPage(ctx, userID, q)
	if err != nil {
		return nil, common.WrapBizError(common.DATABASE_ERROR, "查询收藏列表失败", err)
	}

	list := make([]vo.FavoriteVO, 0, len(rows))
	for _, row := range rows {
		createTime := ""
		if !row.CreatedAt.IsZero() {
			createTime = row.CreatedAt.Format(time.DateTime)
		}
		list = append(list, vo.FavoriteVO{
			ID:         row.ID,
			UserID:     row.UserID,
			TargetType: row.TargetType,
			TargetID:   row.TargetID,
			TargetName: row.AlgorithmName,
			IsInvalid:  row.IsInvalid != 0,
			CreateTime: createTime,
		})
	}

	return &vo.PageResult[vo.FavoriteVO]{List: list, Total: total}, nil
}

func (s *FavoriteService) GetStatus(ctx context.Context, userID int64, targetType string, targetID int64) (*vo.FavoriteStatusVO, error) {
	count, err := s.favRepo.CountByUserAndType(ctx, userID, targetType)
	_ = count // count for debugging if needed
	fav, err := s.favRepo.FindByUserAndTarget(ctx, userID, targetType, targetID)
	if err != nil {
		return nil, common.WrapBizError(common.DATABASE_ERROR, "查询收藏状态失败", err)
	}
	return &vo.FavoriteStatusVO{
		TargetType: targetType,
		TargetID:   targetID,
		Favorited:  fav != nil,
	}, nil
}

func (s *FavoriteService) GetCount(ctx context.Context, userID int64, targetType string) ([]vo.FavoriteCountVO, error) {
	rows, err := s.favRepo.CountGroupByType(ctx, userID, targetType)
	if err != nil {
		return nil, common.WrapBizError(common.DATABASE_ERROR, "查询收藏数量失败", err)
	}

	countMap := make(map[string]int64)
	for _, row := range rows {
		countMap[row.TargetType] = row.Count
	}

	result := make([]vo.FavoriteCountVO, 0, len(validTargetTypes))
	for _, t := range validTargetTypes {
		if targetType != "" && targetType != t {
			continue
		}
		result = append(result, vo.FavoriteCountVO{
			TargetType: t,
			Count:      countMap[t],
		})
	}
	return result, nil
}

func (s *FavoriteService) MarkInvalid(ctx context.Context, targetType string, targetIDs []int64) error {
	if len(targetIDs) == 0 {
		return nil
	}
	return s.favRepo.MarkInvalid(ctx, targetType, targetIDs)
}

// levelCapacityKeys 会员等级 → 收藏容量字典键映射（sys_dict: favorite_capacity）
var levelCapacityKeys = map[string]string{
	"level_0": "default",
	"level_1": "vip1",
	"level_2": "vip2",
	"level_3": "svip",
}

// getCapacity 按用户当前会员等级从字典读取收藏容量上限。
// 缺键/读取失败回退默认容量 200，升级后容量立即扩展（实时读字典）。
func (s *FavoriteService) getCapacity(ctx context.Context, userID int64) int {
	key := "default"
	if member, err := s.memberRepo.FindByUserID(ctx, userID); err == nil && member != nil {
		if k, ok := levelCapacityKeys[member.LevelCode]; ok {
			key = k
		}
	}
	return int(dictservice.GetIntValue(ctx, s.dictSvc, favoriteCapacityDictType, key, 200))
}

var _ IFavoriteService = (*FavoriteService)(nil)

File: dehaze-python/app/service/favorite_service.py
"""
收藏管理 Service

核心业务逻辑：
- 添加收藏（容量校验 + 唯一约束防重 + 取消后重新收藏恢复 deleted=0）
- 批量取消收藏（逻辑删除）
- 收藏列表分页查询
- 收藏状态查询
- 收藏数量统计
- mark_invalid 预留回调
"""

from datetime import datetime

from sqlalchemy.ext.asyncio import AsyncSession

from app.core.code import ResultCode
from app.core.exceptions import BusinessException
from app.repository.favorite_repository import favorite_repository
from app.repository.member_repository import member_repository
from app.service.dict_service import get_dict_int

# 会员等级 → favorite_capacity 字典键映射（对齐收藏管理 §5/§11.1）
LEVEL_TO_CAPACITY_KEY = {
    "level_0": "default",
    "level_1": "vip1",
    "level_2": "vip2",
    "level_3": "svip",
}
# 设计默认容量（与 config/sql/data/sys_dict.sql 种子一致，缺键时回退）
CAPACITY_DEFAULTS = {
    "default": 200,
    "vip1": 500,
    "vip2": 1000,
    "svip": 3000,
}


def _format_dt(dt: datetime | None) -> str | None:
    if dt is None:
        return None
    return dt.strftime("%Y-%m-%d %H:%M:%S")


class FavoriteService:
    async def _get_capacity(self, db: AsyncSession, user_id: int) -> int:
        """根据用户会员等级返回收藏容量上限（sys_dict: favorite_capacity 实时读取）"""
        member = await member_repository.get_by_user_id(db, user_id)
        level_code = member.level_code if member else "level_0"
        key = LEVEL_TO_CAPACITY_KEY.get(level_code, "default")
        return await get_dict_int(db, "favorite_capacity", key, CAPACITY_DEFAULTS.get(key, 200))

    async def add(
        self,
        db: AsyncSession,
        user_id: int,
        target_type: str,
        target_id: int,
    ) -> int:
        """添加收藏

        1. 对象存在性校验（algorithm/dataset/result 必须存在，否则 A0401）
        2. 容量校验
        3. 单条 upsert（冲突时复活 deleted=0, is_invalid=0）
        """
        if not await favorite_repository.target_exists(db, target_type, target_id):
            raise BusinessException(ResultCode.RESOURCE_NOT_FOUND, "收藏目标不存在")

        current_count = await favorite_repository.count_user_favorites(db, user_id)
        capacity = await self._get_capacity(db, user_id)
        if current_count >= capacity:
            raise BusinessException(
                ResultCode.BUSINESS_ERROR,
                f"收藏已达上限（{capacity}条），请清理后重试",
            )

        # upsert：已收藏（deleted=0）幂等返回原 id；已取消（deleted=1）复活返回原 id
        return await favorite_repository.upsert_by_user_and_target(
            db, user_id, target_type, target_id
        )

    async def delete_by_ids(
        self,
        db: AsyncSession,
        user_id: int,
        ids: list[int],
    ) -> None:
        """批量取消收藏（逻辑删除）"""
        if not ids:
            return
        await favorite_repository.soft_delete_by_ids(db, ids, user_id)

    async def get_page(
        self,
        db: AsyncSession,
        user_id: int,
        query: dict,
    ) -> dict:
        """收藏列表分页查询"""
        items, total = await favorite_repository.get_page(
            db,
            user_id,
            query["pageNum"],
            query["pageSize"],
            target_type=query.get("targetType"),
            keywords=query.get("keywords"),
            sort_by=query.get("sortBy"),
            sort_order=query.get("sortOrder"),
        )

        list_data = [
            {
                "id": item["favorite"].id,
                "userId": item["favorite"].user_id,
                "targetType": item["favorite"].target_type,
                "targetId": item["favorite"].target_id,
                "targetName": item.get("target_name"),
                "targetSummary": None,
                "targetThumbnail": item.get("target_thumbnail"),
                "isInvalid": bool(item["favorite"].is_invalid),
                "createTime": _format_dt(item["favorite"].create_time),
            }
            for item in items
        ]
        return {"list": list_data, "total": total}

    async def get_status(
        self,
        db: AsyncSession,
        user_id: int,
        target_type: str,
        target_id: int,
    ) -> dict:
        """检查是否已收藏"""
        existing = await favorite_repository.get_by_user_and_target(
            db, user_id, target_type, target_id
        )
        return {
            "targetType": target_type,
            "targetId": target_id,
            "favorited": existing is not None,
        }

    async def get_count(
        self,
        db: AsyncSession,
        user_id: int,
        target_type: str | None = None,
    ) -> list[dict]:
        """收藏数量统计（按类型分组）"""
        counts = await favorite_repository.get_count_by_type(db, user_id, target_type)
        return [{"targetType": c["target_type"], "count": c["count"]} for c in counts]


favorite_service = FavoriteService()

